runtime/interpreter/value.go

```diff
@@ -12939,6 +12939,558 @@ func (Word64Value) ChildStorables() []atree.Storable {
+		buffer = make([]byte, length)
+	}
+
+	buffer[0] = byte(HashInputTypeWord128)
+	copy(buffer[1:], b)
+	return buffer
+}
+
+func ConvertWord128(memoryGauge common.MemoryGauge, value Value, locationRange LocationRange) Value {
+	return NewWord128ValueFromBigInt(
+		memoryGauge,
+		func() *big.Int {
+
+			var v *big.Int
+
+			switch value := value.(type) {
+			case BigNumberValue:
+				v = value.ToBigInt(memoryGauge)
+
+			case NumberValue:
+				v = big.NewInt(int64(value.ToInt(locationRange)))
+
+			default:
+				panic(errors.NewUnreachableError())
+			}
+
+			if v.Cmp(sema.Word128TypeMaxIntBig) > 0 || v.Sign() < 0 {
+				// When Sign() < 0, Mod will add sema.Word128TypeMaxIntPlusOneBig
+				// to ensure the range is [0, sema.Word128TypeMaxIntPlusOneBig)
+				v.Mod(v, sema.Word128TypeMaxIntPlusOneBig)
+			}
+
+			return v
+		},
+	)
+}
+
+func (v Word128Value) BitwiseOr(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue {
+	o, ok := other.(Word128Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationBitwiseOr,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	return NewWord128ValueFromBigInt(
+		interpreter,
+		func() *big.Int {
+			res := new(big.Int)
+			return res.Or(v.BigInt, o.BigInt)
+		},
+	)
+}
+
+func (v Word128Value) BitwiseXor(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue {
+	o, ok := other.(Word128Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationBitwiseXor,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	return NewWord128ValueFromBigInt(
+		interpreter,
+		func() *big.Int {
+			res := new(big.Int)
+			return res.Xor(v.BigInt, o.BigInt)
+		},
+	)
+}
+
+func (v Word128Value) BitwiseAnd(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue {
+	o, ok := other.(Word128Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationBitwiseAnd,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	return NewWord128ValueFromBigInt(
+		interpreter,
+		func() *big.Int {
+			res := new(big.Int)
+			return res.And(v.BigInt, o.BigInt)
+		},
+	)
+
+}
+
+func (v Word128Value) BitwiseLeftShift(interpreter *Interpreter, other IntegerValue, locationRange LocationRange) IntegerValue {
+	o, ok := other.(Word128Value)
+	if !ok {
+		panic(InvalidOperandsError{
+			Operation: ast.OperationBitwiseLeftShift,
+			LeftType:  v.StaticType(interpreter),
+			RightType: other.StaticType(interpreter),
+		})
+	}
+
+	return NewWord128ValueFromBigInt(
+		interpreter,
+		func() *big.Int {
+			res := new(big.Int)
+			if o.BigInt.Sign() < 0 {
+				panic(UnderflowError{LocationRange: locationRange})
+			}
+			if !o.BigInt.IsUint64() {
+				panic(OverflowError{LocationRange: locationRange})
+			}
+			return res.Lsh(v.BigInt, uint(o.BigInt.Uint64()))
```

Comment:
The `Lsh` and `Rsh` functions from the big integer module only supports uint64 for the shift size. So I am overflowing here for Word128. Not sure about the alternatives here. May be we can document that `a << b` is only supported when `b` fits a `uint64`.

Reply:
Yes, probably best to document this

Reply:
Sure, will do that in the documentation PR.